crates/ruma-client-api/src/error.rs

```diff
@@ -230,6 +230,31 @@ impl fmt::Display for Error {
 
 impl std::error::Error for Error {}
 
+impl From<FromHttpRequestError> for Error {
+    fn from(error: FromHttpRequestError) -> Self {
+        let kind = match error {
+            FromHttpRequestError::Deserialization(DeserializationError::Json(t)) => {
+                match t.is_syntax() {
+                    true => ErrorKind::NotJson,
+                    false => ErrorKind::BadJson,
+                }
+            }
+            FromHttpRequestError::Deserialization(_anything) => ErrorKind::InvalidParam,
+            _ => ErrorKind::Unknown,
+        };
+        Self {
+            message: match &kind {
+                ErrorKind::NotJson => "The JSON is malformed and cannot be parsed".into(),
```

Comment:
By splitting this function into two `match`es, you can't include the underlying error message in the output `message`. I would suggest combining them and using a message like

```suggestion
                ErrorKind::NotJson => {
                    format!("The request body is not valid JSON: {}", json_error),
```